src/vfs/virtualFileSystemWithWatch.ts

```diff
@@ -0,0 +1,865 @@
+/* @internal */
+namespace ts.VirtualFS {
+    export interface VirtualServerHostCreationParameters {
+        useCaseSensitiveFileNames?: boolean;
+        executingFilePath: string;
+        currentDirectory?: string;
+        newLine?: string;
+        windowsStyleRoot?: string;
+        environmentVariables?: ESMap<string, string>;
+        runWithoutRecursiveWatches?: boolean;
+        runWithFallbackPolling?: boolean;
+    }
+
+    export function createVirtualWatchedSystem(fileOrFolderList: readonly FileOrFolderOrSymLink[], params?: VirtualServerHostCreationParameters): VirtualServerHost {
+        return new VirtualServerHost(fileOrFolderList, params);
+    }
+
+    export function createVirtualServerHost(fileOrFolderList: readonly FileOrFolderOrSymLink[], params?: VirtualServerHostCreationParameters): VirtualServerHost {
+        const host = new VirtualServerHost(fileOrFolderList, params);
+        // Just like sys, patch the host to use writeFile
+        patchWriteFileEnsuringDirectory(host);
+        return host;
+    }

```

Comment:
I think as part of this you should just provide constructor here and test do what they were doing ?

Reply:
I'm not sure what you mean:
provide a constructor for something besides a VirtualServerHost? Maybe a TestServerHost? Or do you mean that tests should call `new VirtualServerHost` and patch it themselves?